### ISSUE TITLE
Rule Proposal: no-multi-code-point-character-in-character-class

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#10049]
**Please describe what the rule should do:**

If RegExp character class has multi-code-point character, it causes unexpected behavior, even if you are using `u` flag.

```js
/^[❇️]$/u.test("❇️") //→ false
```

In above example, the `❇️` is made with two code points U+2747 and U+FE0F. JavaScript doesn't handle it as a character, so the character class matches to U+2747 or U+FE0F rather than the character. The U+FE00-FE0F is [Variation Selectors](https://en.wikipedia.org/wiki/Variation_Selectors_(Unicode_block)).

Another well-known case is ZWJ U+200D. ZWJ joins multiple code points to a character.

```js
/^[👨‍👩‍👦]$/u.test("👨‍👩‍👦") //→ false
```

So, this rule prevents characters which are made with multiple code points in character class syntax.

**What category of rule is this? (place an "X" next to just one item)**

[X] Warns about a potential error

**Provide 2-3 code examples that this rule will warn about:**

```js
/[👍]/ // ERROR: a surrogate pair in a character class without u flag.
/[❇️]/u // ERROR: variation selectors in a character class.
/[👨‍👩‍👦]/u // ERROR: ZWJ in a character class.
/[🇯🇵]/u // ERROR: a U+1F1E6-1F1FF pair in a character class.
/[👶🏻]/u // ERROR: an emoji which is made with an emoji and skin tone selector, in a character class.
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This is not library specific matter and it's difficult to distinguish whether character class syntax handles the character as one character or not.

**Comment by not-an-aardvark** [comment]
:+1:, but I feel like the rule name is a bit too long. How about something like `no-invalid-character-classes` or `no-surrogate-pair-character-classes`?

**Comment by mysticatea** [comment]
I agree. In fact, my first idea was `no-destructuring-character-classes`, but the word is used in destructuring assignment syntax.

`no-invalid-character-classes` is not bad.
`no-surrogate-pair-character-classes` sounds misleading since `u` flag can handle surrogate pairs correctly.

Hmm, how about `no-dismantling-character-classes`? The purpose of the rule is to report character classes which dismantle one character to multiple characters.

**Comment by platinumazure** [comment]
Marking as accepted.

I've also added "needs bikeshedding" since we need to agree on a name.
